Marlin/src/core/types.h

```diff
@@ -0,0 +1,156 @@
+/**
+ * Marlin 3D Printer Firmware
+ * Copyright (c) 2019 MarlinFirmware [https://github.com/MarlinFirmware/Marlin]
+ *
+ * Based on Sprinter and grbl.
+ * Copyright (c) 2011 Camiel Gubbels / Erik van der Zalm
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+#pragma once
+
+#include <stdint.h>
+
+#ifndef UNUSED
+  #define UNUSED(x) ((void)(x))
+#endif
+
+typedef uint32_t millis_t;
+
+#define PENDING(NOW,SOON) ((int32_t)(NOW-(SOON))<0)
+#define ELAPSED(NOW,SOON) (!PENDING(NOW,SOON))
+
+// Position types
+template<typename T> struct XYZEVal;
+template<typename T> struct XYZVal;
+
+template<typename T>
+struct XYVal {
+  union {
+    struct { T x, y; };
+    T pos[2];
+  };
+  void set(const T px, const T py, const T pz=0, const T pe=0) { x = px; y = py; UNUSED(pz); UNUSED(pe); }
+  void reset() { x = y = 0; }
+  XYVal<T>& operator=(const XYZVal<T>& xyz) { set(xyz.x, xyz.y); return *this; }
+  XYVal<T>& operator=(const XYZEVal<T>& xyze) { set(xyze.x, xyze.y); return *this; }
+  XYVal<T> operator+(const XYVal<T>& rs) { XYVal<T> ls = *this; ls.x += rs.x; ls.y += rs.y; return ls; }
+  XYVal<T> operator-(const XYVal<T>& rs) { XYVal<T> ls = *this; ls.x -= rs.x; ls.y -= rs.y; return ls; }
+  XYVal<T>& operator+=(const XYVal<T>& rs) { x += rs.x; y += rs.y; return *this; }
+  XYVal<T>& operator-=(const XYVal<T>& rs) { x -= rs.x; y -= rs.y; return *this; }
+  T& operator[](int i) { return pos[i]; }
```

Comment:
Put this...
T& operator[](size_t i)             { return pos[i]; }
const T& operator[](size_t i) const { return pos[i]; }

Reply:
It's worth another try. The compiler was barking at methods marked `const` in this way, so I have been leaving it out.

Reply:
Ah, right. I had to deal with this when I wrote a custom string class for "FretPet." The `const` method is used when the data is `const`. Otherwise the non-`const` version is used. I also ran across some issues when declaring `virtual` instances. I will have to revisit that and see if a similar solution applies. For now I just declared a type whose members are `virtual`.